internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+	"time"
+)
+
+type Semaphore interface {
+	//Acquire a slot in the semaphore with a timeout
+	Acquire(timeout time.Duration) bool
+
+	//Release a slot in the semaphore.
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
+
+type semaphore struct {
+	slots chan struct{}
+}
+
+func NewSemaphore(limit int) Semaphore {
+	if limit < 0 {
+		log.Panic("E! Semaphore size limit must not be negative")
+	}
+
+	sem := &semaphore{
+		slots: make(chan struct{}, limit),
+	}
+
+	for i := 0; i < cap(sem.slots); i++ {
+		sem.slots <- struct{}{}
+	}
+
+	return sem
+}
+
+func (sem *semaphore) Acquire(timeout time.Duration) bool {
+	if (timeout > 0) {
+		timer := time.NewTimer(timeout)
+		defer timer.Stop()
+		select {
+		case <-sem.slots:
+			return true
+		case <-timer.C:
+		}
+	}
```

Comment:
We don't need to have a timer here. We can just use `time.After()` to fail through:
```suggestion
	if (timeout > 0) {
		select {
		case <-sem.slots:
			return true
		case <-time.After(timeout):
		}
	}
```